spec/index.md

```diff
@@ -0,0 +1,308 @@
+
+> **<sup>Lexer</sup>**\
+> KW_ABSTRACT       : `abstract`\
+> KW_ABSTRACT       : `async`\
+> KW_ABSTRACT       : `await`\
+> KW_DO             : `do`\
+> KW_FINAL          : `final`\
+> KW_IMPL           : `impl`\
+> KW_MACRO          : `macro`\
+> KW_MATCH          : `match`\
+> KW_MUT            : `mut`\
+> KW_OVERRIDE       : `override`\
+> KW_STATIC         : `static`\
+> KW_TRAIT          : `trait`\
+> KW_TYPE           : `type`\
+> KW_TYPEOF         : `typeof`\
+> KW_USE            : `use`\
+> KW_VIRTUAL        : `virtual`\
+> KW_WHERE          : `where`\
+> KW_YIELD          : `yield`
+
+
+### 1.2. Identifiers
+
+> **<sup>Lexer:<sup>**\
+> IDENTIFIER_OR_KEYWORD :\
+> &nbsp;&nbsp; &nbsp;&nbsp; [`a`-`z` `A`-`Z`]&nbsp;[`a`-`z` `A`-`Z` `0`-`9` `_`]<sup>\*</sup>\
+> &nbsp;&nbsp; | `_` [`a`-`z` `A`-`Z` `0`-`9` `_`]<sup>+</sup>
+><sub>*Except a [strict] or [reserved] keyword*</sub>
+>
+
+An identifier is any nonempty ASCII string of the following form:
+
+Either
+
+* The first character is a letter.
+* The remaining characters are alphanumeric or `_`.
+
+Or
+
+* The first character is `_`.
+* The identifier is more than one character. `_` alone is not an identifier.
+* The remaining characters are alphanumeric or `_`.
+
+[strict]: index.md#strict-keywords
+[reserved]: index.md#reserved-keywords
+
+
+## 3. Items
+
+### 3.1. Functions
+
+> **<sup>Syntax</sup>**\
+> _Function_ :\
+> &nbsp;&nbsp; _FunctionDecorator_ `@`[IDENTIFIER]<sup>?</sup>\
+> &nbsp;&nbsp; `def` [IDENTIFIER]\
+> &nbsp;&nbsp; &nbsp;&nbsp; `(` _FunctionParameters_<sup>?</sup> `)`\
+> &nbsp;&nbsp; &nbsp;&nbsp; _FunctionReturnType_<sup>?</sup>\
+> &nbsp;&nbsp; &nbsp;&nbsp; _FunctionHeaderTerminator_<sup>?</sup>\
+> &nbsp;&nbsp; &nbsp;&nbsp; [_BlockExpression_]
+>
+>
+> _FunctionDecorators_ :\
+> &nbsp;&nbsp; _FunctionDecorator_<sup>\*</sup>
+>
+> _FunctionDecorator_ :\
+> &nbsp;&nbsp; `@`[IDENTIFIER]
+>
+> _FunctionParameters_ :\
+> &nbsp;&nbsp; _FunctionParam_ (`,` _FunctionParam_)<sup>\*</sup> `,`<sup>?</sup>
+>
+> _FunctionParam_ :\
+> &nbsp;&nbsp; [_Pattern_] `:` [_Type_]
+>
+> _FunctionReturnType_ :\
+> &nbsp;&nbsp; `->` [_Type_]
+
+> _FunctionHeaderTerminator_ :\
+> &nbsp;&nbsp; `:`
+
+A _function_ consists of a [block], along with a name and a set of parameters.
+Other than a name, all these are optional. Functions are declared with the
+keyword `def`. Functions may declare a set of *input* [*variables*][variables]
+as parameters, through which the caller passes arguments into the function, and
+the *output* [*type*][type] of the value the function will return to its caller
+on completion.
+
+When referred to, a _function_ yields a first-class *value* of the
+corresponding zero-sized [*function item type*], which
+when called evaluates to a direct call to the function.
+
+A function header ends with a colon (`:`) after which the function body begins.
+
+For example, this is a simple function:
+
+```python
+def answer_to_life_the_universe_and_everything() -> u32:
```

Comment:
Yeah, I can change that to `u256` but I think we eventually want support for all multiples of 8 up to 256, right?

Reply:
Yeah, we should have `u8`, `u16`, `u32`, ..., `u256`. No need to change, was just double checking since some type systems use bytes for sizing.

Reply:
So I just noticed that the test contracts have `uint256` instead of `u256`... This is something that isn't specified anywhere, even in the grammar. I would be open to changing this if you think `u256` is cleaner, as it's a fairly trivial change.

Reply:
Sounds good. I'm :+1: using `uN`  and `iN` mainly because it saves some space without compromising on readability and is inline with rust.